Deprecate manually vectorized methods in favor of dot syntax, monolithic edition

Sacha0 opened the pull request:
This pull request deprecates almost all remaining manually vectorized methods, incorporating #19713, #19710, #19709, #18609, #18608, #18607, #18593, #18590, #18586, #18576, #18575, #18571, #18564, and #18512 into a single pull request. The last commit reinstates a few `&`, `|`, and `xor` methods over `BitArray`s as `broadcast` specializations (in accord with discussion in the corresponding pull requests).

(I dropped #19712 from this set: #19712 turned out to need a fair bit more work and touches a lot of code. Best done as a separate pull request if consensus favors deprecating those methods (two-arg vectorized `complex`); I *might* be able to put that together tomorrow if so.)

Best!

Sacha0 committed:
Deprecate manually vectorized big methods in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate manually vectorized abs2 methods in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate manually vectorized methods for trigonometric and hyperbolic functions in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate manually vectorized trunc methods in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate manually vectorized float methods in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate manually vectorized ceil methods in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate manually vectorized round methods in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate manually vectorized mod methods in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate manually vectorized div methods in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate manually vectorized rem methods in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate manually vectorized clamp methods in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized & in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized | in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized min and max over pairs of sparse matrices.

Sacha0 committed:
Restore a few methods for &, |, and xor over BitArrays as broadcast specializations.

tkelman commented on test/sparse/sparsevector.jl:
```diff
@@ -1,4 +1,4 @@
-# This file is a part of Julia. License is MIT: http://julialang.org/license
+<# This file is a part of Julia. License is MIT: http://julialang.org/license
```
this shouldn't be here

Sacha0 replied:
Top-of-file gremlin strikes again. Fixed. Thanks!

tkelman replied:
https://github.com/JuliaLang/julia/pull/19721 also means rebase needed for end-of-deprecated.jl

Sacha0 replied:
Rebased. Thanks!

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -1405,7 +1405,6 @@ conj!(A::SparseMatrixCSC) = (broadcast!(conj, A.nzval, A.nzval); A)
 
 # TODO: The following definitions should be deprecated.
 ceil{To}(::Type{To}, A::SparseMatrixCSC) = ceil.(To, A)
-floor{To}(::Type{To}, A::SparseMatrixCSC) = floor.(To, A)
```
had this been missing a `To<:Integer` constraint?

Sacha0 replied:
Depending on how you interpret the docstring? `floor(T, x) converts the result to type T, throwing an InexactError if the value is not representable.`

tkelman replied:
The existing vectorized methods on a lot of these rounding functions were a bit inconsistent on whether or not they constrained the type argument to Integer subtypes. Is that just a matter of when you want your MethodError, or does it make a real difference?

Sacha0 replied:
Good question. I thought perhaps the docstring implies that e.g. `floor(Float32, 1.7) == 1.0f0` should yield `true` (though `floor(Float32, 1.7)` presently fails), and hence the absence of `<:Integer` constraints. Or is the docstring itself too loose? Thanks!

tkelman replied:
I dunno, but I guess the deprecations may as well leave off the Integer constraint and hand off the problem to the scalar method, seems safer than an incomplete deprecation

Sacha0 replied:
Sounds good :).

tkelman commented:
This will likely show big regressions for any of the benchmarks that are using now-deprecated syntax. I'm more curious about whether it shows any medium regressions for any refactorings of the internals here that might be using different code paths than they used to be: @nanosoldier `runbenchmarks(ALL, vs = ":master")`

(also question for @jrevels, does nanosoldier use a different branch of basebenchmarks for testing against different branches of Julia?)

tkelman commented:
There are ambiguities and a use of one of the deprecated functions in a test. It would be good to squash the fixes for these into the specific commit that introduced them if you can, since the sequence of 15 commits here are logically separated and make more sense not-squashed - as long as the build passes at each intermediate steps, to avoid accidentally breaking future bisects.

Sacha0 committed:
Deprecate a few remaining vectorized methods over SparseVectors. (And associated cleanup.)

Sacha0 committed:
Replace former error path tests for vectorized &, |, xor, min, and max over SparseMatrixCSCs with similar tests of sparse map.

Sacha0 commented:
Fixed the ambiguities introduced by rebasing post #19690-merge (and in the appropriate commits). With benefit of #19690, added a separate commit that removes a few vectorized methods over `SparseVector`s and performs minor associated cleanup. Also added a commit replacing removed error-path tests for deprecated vectorized `&`, `|`, `xor`, `min`, and `max` methods over `SparseMatrixCSC`s with similar tests of sparse `map` as discussed in https://github.com/JuliaLang/julia/pull/19713#discussion_r94269292. Thanks again!

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19791#issuecomment-269850951) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/9f35457dae975c90e26cb05b52be80ea4f0fc23a/94f7b60_vs_26c8d85/report.md). cc @jrevels

tkelman commented:
That's less change than I expected. string join is often noisy, but `["array","comprehension",("collect","Array{Float64,1}")]` might be real? Let's go again on the updated version @nanosoldier `runbenchmarks(ALL, vs = ":master")`

tkelman reviewed (approved):
addressed my own review comments in https://github.com/Sacha0/julia/pull/1 - any perf regressions can hopefully be fixed after feature freeze

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19791#issuecomment-269857635) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/7eb72c9e1653485630c558953980fb8e4e67a732/267d9e5_vs_26c8d85/report.md). cc @jrevels

tkelman commented:
This is such a conflict-magnet that I'm going to merge now and open my trivial little PR against master instead.

jrevels commented:
> (also question for @jrevels, does nanosoldier use a different branch of basebenchmarks for testing against different branches of Julia?)

No, all versions of Julia benchmark against the `nanosoldier` branch.

Sacha0 commented:
Tony, you're amazing. Do you ever sleep, or instead just recover in blocks of hypnagogic PR review? Thanks for your tremendous efforts!

Sacha0 commented:
> I'm curious how the new general sparse vector broadcast code compares in a benchmark against this older method for abs.(v). (Don't need to spend time checking right now, can wait.)

Likewise. I doubt generic sparse broadcast matches the performance of specializations like that. Making a note to benchmark this case when time allows. Thanks!

giordano commented:
Going from the old manually vectorized method of the 2-arg variant of `trunc` to the dot syntax, I experience a large performance regression.

Before this PR (manually vectorized method):
```julia
julia> versioninfo()
Julia Version 0.6.0-dev.1843
Commit c38a5a304 (2017-01-01 14:09 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i7-4700MQ CPU @ 2.40GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, haswell)

julia> using BenchmarkTools

julia> arr = collect(0.0:0.1:10.0);

julia> @benchmark trunc(Int, arr)
BenchmarkTools.Trial: 
  memory estimate:  912.00 bytes
  allocs estimate:  2
  --------------
  minimum time:     368.604 ns (0.00% GC)
  median time:      379.300 ns (0.00% GC)
  mean time:        424.871 ns (6.22% GC)
  maximum time:     4.681 μs (83.10% GC)
  --------------
  samples:          10000
  evals/sample:     207
  time tolerance:   5.00%
  memory tolerance: 1.00%
```
after the PR (dot syntax):
```julia
julia> versioninfo()
Julia Version 0.6.0-dev.1807
Commit 26c8d856a (2016-12-31 04:12 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i7-4700MQ CPU @ 2.40GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, haswell)

julia> using BenchmarkTools

julia> arr = collect(0.0:0.1:10.0);

julia> @benchmark trunc.(Int, arr)
BenchmarkTools.Trial: 
  memory estimate:  3.61 kb
  allocs estimate:  132
  --------------
  minimum time:     46.124 μs (0.00% GC)
  median time:      47.363 μs (0.00% GC)
  mean time:        48.439 μs (0.74% GC)
  maximum time:     3.685 ms (97.86% GC)
  --------------
  samples:          10000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

Can someone please confirm?

KristofferC commented:
Make sure to not benchmark in global scope and use `const` but even so, there is a significant regression.

giordano commented:
Thanks, I updated my benchmarks with the use of `const`, `trunc` is still 100× slower after this PR.  I'm going open a new issue.

tkelman commented:
We somehow didn't see this in any tests, but ref https://discourse.julialang.org/t/is-0-a-1-going-away/1313/2 - chained comparisons like `A .< B .< C` expand to `(A .< B) & (B .< C)`, which now throws a depwarn if any are arrays. Would it be always equivalent to the former behavior if we made it instead use `.&` in the expansion, or would it have to depend on the types? cc @stevengj

stevengj commented:
It should be equivalent to the former behavior if we made it use `.&`.

Sacha0 commented:
> I'm curious how the new general sparse vector broadcast code compares in a benchmark against this older method for abs.(v). (Don't need to spend time checking right now, can wait.)

@tkelman, generic sparse `broadcast` performs surprisingly well in this case:
```julia
julia> using BenchmarkTools

julia> begin
           vecabs(v::SparseVector) = SparseVector(v.n, copy(v.nzind), abs.(v.nzval))

           xs = sprand(10^2, 0.2);
           ds = @benchmark abs.($xs);
           vs = @benchmark vecabs($xs);
           println(judge(median(ds), median(vs)))

           xm = sprand(10^4, 0.1);
           dm = @benchmark abs.($xm);
           vm = @benchmark vecabs($xm);
           println(judge(median(dm), median(vm)))

           xl = sprand(10^6, 0.1);
           dl = @benchmark abs.($xl);
           vl = @benchmark vecabs($xl);
           println(judge(median(dl), median(vl)))
       end

BenchmarkTools.TrialJudgement:
  time:   +44.44% => regression (5.00% tolerance)
  memory: +0.00% => invariant (1.00% tolerance)
BenchmarkTools.TrialJudgement:
  time:   +43.54% => regression (5.00% tolerance)
  memory: +0.00% => invariant (1.00% tolerance)
BenchmarkTools.TrialJudgement:
  time:   +27.62% => regression (5.00% tolerance)
  memory: +0.00% => invariant (1.00% tolerance)
```
Eking out a bit more performance might be possible (e.g. with a few `@propagate_inbounds` decorations). Best!

tkelman commented:
Cool. I'd add the one line of code specialization (despite the fact it'll be circumvented in fusing cases) if cleverer tweaks to the general case can't get within 10-20ish percent, but it's not that bad.

Sacha0 commented:
Note that the above comparison isn't apples-to-apples: sparse `map[!]/broadcast[!]` avoids storing numerical zeros, whereas the specialization above does not. (The above performance difference is in line with that observed between the former `nz2z` and `nz2nz` sparse `broadcast` classes for similar cases, as benchmarked/discussed elsewhere.) Best!
